### PR TITLE
storage: expand godoc for baseQueue

### PR DESCRIPTION
**Pull request by danhhz**
I had to remind myself how this all works while thinking about learner
replicas, so adding my notes for the benefit of others.

Release note: None

### CODE REVIEW

**Review by tbg — APPROVED** [review]
Thanks!

> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37949#-:-LgXLqPC4wJELwIPulRA:b-ucfknh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz)

___
*[pkg/storage/queue.go, line 296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37949#-LgXLTSnDLBqpaJFYxxP:-LgXLTSo3gi890JF5P2g:b-dvl6sj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7126474adf891de3b73c0967fd5c59a76a142b3/pkg/storage/queue.go#L296)):*
> ```Go
> //
> // Replicas are typically added asynchronously through `MaybeAddAsync` or
> // `AddAsync`, though this can also be done synchronously though `maybeAdd` or
> ```

We don't want to use the internal methods since they require hard thinking about lock ordering. If the queues were in their own package, this method wouldn't be accessible, and I'd treat it as such in the documentation, perhaps clarifying why there isn't a way to add to a queue synchronously.

**Review by danhhz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/storage/queue.go, line 296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37949#-LgXLTSnDLBqpaJFYxxP:-LgXwFnvCQIfcseQSWbf:b-k2z51m) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7126474adf891de3b73c0967fd5c59a76a142b3/pkg/storage/queue.go#L296)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

We don't want to use the internal methods since they require hard thinking about lock ordering. If the queues were in their own package, this method wouldn't be accessible, and I'd treat it as such in the documentation, perhaps clarifying why there isn't a way to add to a queue synchronously.
</blockquote></details>

Okay! And the reason that we don't have a way to add to a queue synchronously is because it would require hard thinking about lock ordering? I don't quite see that. Which locks?

**Review by tbg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz)

___
*[pkg/storage/queue.go, line 296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37949#-LgXLTSnDLBqpaJFYxxP:-LgYSTty1HhB89oq2lVm:bm752yk) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7126474adf891de3b73c0967fd5c59a76a142b3/pkg/storage/queue.go#L296)):*
<details><summary><i>Previously, danhhz (Daniel Harrison) wrote…</i></summary><blockquote>

Okay! And the reason that we don't have a way to add to a queue synchronously is because it would require hard thinking about lock ordering? I don't quite see that. Which locks?
</blockquote></details>

See https://github.com/cockroachdb/cockroach/issues/36413

TL;DR: `Add` calls `shouldQueue` which can then take various replica locks, so the caller of `Add` has to be aware of all that in relation the the locks it's holding itself. That seems like a frail contract, and queues are async anyway without any guarantee that what you add will eventually be processed, so the exclusive-async contract just forces you to realize this early.

**Review by danhhz — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/storage/queue.go, line 296 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/37949#-LgXLTSnDLBqpaJFYxxP:-LgYcZ7l5fepOuxf-fdI:b-yjop50) ([raw file](https://github.com/cockroachdb/cockroach/blob/b7126474adf891de3b73c0967fd5c59a76a142b3/pkg/storage/queue.go#L296)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

See https://github.com/cockroachdb/cockroach/issues/36413

TL;DR: `Add` calls `shouldQueue` which can then take various replica locks, so the caller of `Add` has to be aware of all that in relation the the locks it's holding itself. That seems like a frail contract, and queues are async anyway without any guarantee that what you add will eventually be processed, so the exclusive-async contract just forces you to realize this early.
</blockquote></details>

Done. Thanks for the context!

I added a note to the docs for MaybeAddCallback after reading the code. Can you double check that my understanding is correct?

**Review by tbg — APPROVED** [review]
> Reviewed 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37949)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @danhhz and @tbg)

___
*[pkg/storage/queue.go, line 705 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37949#-LgZ3vXbBn1FgzrWSHjq:-LgZ3vXbBn1FgzrWSHjr:br5n7p4) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e12593bec074b985c32e7ac6518a096fb731654/pkg/storage/queue.go#L705)):*
> ```Go
> // purgatory, the callback is called immediately with the purgatory error. If
> // the range is not in the queue (either waiting or processing), the method
> // returns false. If the replica this attaches to is dropped from an overfull
> ```

Looks right. We only exec the callbacks if we make it into `finishProcessingReplica`. This behavior is fine for the sole user

https://github.com/cockroachdb/cockroach/blob/7247479317d26d2a6f1b9cd5ee558303534f3da1/pkg/storage/replica_backpressure.go#L120

but we should discourage anyone else from using this. If we need this more, we'll have to fix the semantics, but I know from experience fixing related semantics (fallout from the locking story) that this code doesn't lend itself readily to upholding invariants.

**Review by danhhz — COMMENTED** [review]
Thanks for the review!

bors r=tbg

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/37949#-:-Lgc2b783QDz1LOULUWQ:b-6rrddf)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/storage/queue.go, line 705 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/37949#-LgZ3vXbBn1FgzrWSHjq:-Lgc2TPnDDr2VD3KS-Of:b-7gptwp) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e12593bec074b985c32e7ac6518a096fb731654/pkg/storage/queue.go#L705)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

Looks right. We only exec the callbacks if we make it into `finishProcessingReplica`. This behavior is fine for the sole user

https://github.com/cockroachdb/cockroach/blob/7247479317d26d2a6f1b9cd5ee558303534f3da1/pkg/storage/replica_backpressure.go#L120

but we should discourage anyone else from using this. If we need this more, we'll have to fix the semantics, but I know from experience fixing related semantics (fallout from the locking story) that this code doesn't lend itself readily to upholding invariants.
</blockquote></details>

Done and removed the reference above
